ext/MitsubaLoader/CMitsubaLoader.h

```diff
@@ -49,269 +49,165 @@ namespace bsdf
-		//	otherwise alpha_v[0] is constant alpha_v
+		//	otherwise alpha_v[0] is constant single-float alpha_v
 		//otherwise (different NDF)
 		//	if flag decides to use alpha texture, alpha_u[0..1] is bindless texture ID
-		//	otherwise alpha_u[0] is constant alpha
-		float alpha_u[2];
-		float alpha_v[2];
-		//if flag decides to use spec refl texture, specReflectance[0..1] is bindless texture ID
-		//otherwise specReflectance[0..2] is constant spec refl
-		float specReflectance[3];
-		float dummy1;
-		//if flag decides to use spec transmitance texture, specTransmittance[0..1] is bindless texture ID
-		//otherwise specTransmittance[0..2] is constant spec transmitance
-		float specTransmittance[3];
+		//	otherwise alpha_u[0] is constant single-float alpha
+		uint32_t alpha_u[2];
+		uint32_t alpha_v[2];
+		padding_t<7, max_bsdf_struct_size> padding;
 	} PACK_STRUCT;
-	/**
-	GLSL:
-	struct {
-		uint bitfields;
-		float ext_ior;
-		vec2 alpha_u;
-		vec2 alpha_v;
-		layout (column_major) mat2x3 ior;
-		vec3 specReflectance;
-	};
-	*/
 	struct alignas(16) SAllConductor
 	{
 		uint32_t bitfields;
-		float ext_ior;
+		//actually it's float
+		uint32_t ext_ior;
 		//same as for SAllDielectric::alpha_u,alpha_v
-		float alpha_u[2];
-		float alpha_v[2];
-		float dummy[2];
-		//column-major matrix 3x2 (GLSL's 2x3)
-		//ior[0..2] is column 0 (real part of IoR)
-		//ior[4..6] is column 1 (imaginary part of IoR)
-		float ior[8];
-		//same as SAllDielectric::specReflectance
-		float specReflectance[3];
+		uint32_t alpha_u[2];
+		uint32_t alpha_v[2];
+		//ior[0] real part of IoR in RGB19E7 format
+		//ior[4..6] is imaginary part of IoR in RGB19E7 format
+		uint64_t ior[2];
+		padding_t<8, max_bsdf_struct_size> padding;
 	} PACK_STRUCT;
-	/**
-	GLSL:
-	struct {
-		uint bitfields;
-		vec2 int_ext_ior;
-		vec2 alpha_u;
-		vec2 alpha_v;
-		vec3 specReflectance;
-		vec3 specTransmittance;
-		vec3 diffReflectance;
-	};
-	*/
 	struct alignas(16) SAllPlastic
 	{
 		uint32_t bitfields;
-		float dummy0;
 		//same as SAllDielectric::int_ext_ior
-		float int_ext_ior[2];
+		uint32_t int_ext_ior[2];
 		//same as for SAllDielectric::alpha_u,alpha_v
-		float alpha_u[2];
-		float alpha_v[2];
-		//same as SAllDielectric::specReflectance
-		float specReflectance[3];
-		float dummy1;
-		//same as SAllDielectric::specTransmittance
-		float specTransmittance[3];
-		float dummy2;
-		//if flag decides to use diffuse refl texture, diffReflectance[0..1] is bindless texture ID
-		//otherwise diffReflectance[0..2] is constant diffuse refl
-		float diffReflectance[3];
+		uint32_t alpha_u[2];
+		uint32_t alpha_v[2];
+		padding_t<7, max_bsdf_struct_size> padding;
 	} PACK_STRUCT;
-	/**
-	GLSL:
-	struct {
-		uint bitfields;
-		float thickness;
-		vec2 int_ext_ior;
-		vec2 alpha_u;
-		vec2 alpha_v;
-		vec3 specReflectance;
-		vec3 specTransmittance;
-		vec3 sigmaA;
-	};
-	*/
 	struct alignas(16) SAllCoating
 	{
 		//child index (into BSDF buffer) is stored on 15 highest bits
```

Comment:
yeah, about this, for efficient computation it would make more sense to:

1) Find unique trees (different roots)
2) For every tree, do post-order traversal on the CPU to figure out the order in which BSDF nodes would get evaluated

For establishing the normal to use for shading:
3) error out if more than one (can be the same over and over) `bumpmap` bsdf node gets used in the tree (we don't want to be wasting time on that)
4) record normal tangent frames (max 2, and first one is the smooth non displaced one) to be used for shading and evaluate them at the start of the `bsdf_` functions (all the angles and cosines)
5) for every node in the BSDF tree tag which shading normal tangent frame its going to be using (only need 1 bit).

For the `bsdf_sample` functions:
6) You don't need a stack, every single time you hit a blend BSDF node, you will pick the left or right child according to the `sample.x` and the proportion of the left to right child blend weight, then you will sample the child with `sample.x = (sample.x-pre_cdf)/chosen_pdf; sample.xy = sample.yx;` (rescaled random numbers) and bump the output value by `1.0/chosen_pdf` iteratively until you hit a leaf node.

For the bsdf_eval function:
7) when popping the BSDF node from the post-order traversal stack in hte CPU offline preprocessing step, find out the max-stack-size (for all trees, so you can only define that value via specialization constant after you've constructed all the BSDF trees) for the UberShader (with assumption that a parent Blend node moves the resultant value onto the stack replacing the leftmost child).
8) For every tree, record the nodes in the order they would have gotten **popped off the stack** and record the offset in the per-invocation-local-stack where they would have stored their result (for the blend node, do not store children indices, but stack-offsets to consume).